### PR TITLE
TF generate refactor - Greedy Search

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR is the first step to refactor TF's `generate()` method similar in spirit to https://discuss.huggingface.co/t/big-generate-refactor/1857 . 

## Motivation

The main reasons for the refactor are:
- Make `generate` more readable and easier to understand. 
- Disentangle different components of `generate`, *e.g.* the model_input preparation, the logits processor creation and application, the sub-generation methods. Each of these components of `generate` should be as independent and extendable as possible. We've seen that such a design greatly helps to foster community contributions to the generate method for PyTorch.
- Make `generate` more generally applicable for future use cases. This mostly concerns multi-modal models, such as speech to text and image to text. 

In addition the final goal is to significantly speed up TF's generate method by making it XLA-compatible. This PR should greatly help at making the `generate()` method compatible with XLA by:
- clearly separating the model inputs preparation part and the auto-regressive part. The auto-regressive part is handled in the sub-generation method which should now be much shorter. This is the part which will require some heavy changes to work with XLA. 
- Moving out all of the logits processing logic into a `LogitsProcessorList` class. We can now simply test each of those classes separately whether they are XLA-compatible and don't have to make each of them XLA-compatible. E.g. `bad_token_ids` and `ngram_repeat` are probably not at all XLA-compatible.

## Proposed refactor design and steps to achieve it

TF's `generate` can currently handle more or less **4** different types of generation:
- a) greedy search
- b) sampling (GPT-like generation)
- c) beam search (default use case for translation, etc...)
- d) beam sample (exotic case where sampling is combined with `beam_search`

Currently a) and b) are handled by the sub-method `_generate_no_beam_search` while c) and d) are handled by `_generate_beam_search` .
This is not very readable and understandable. I propose to remove those sub-methods and instead define all four generation methods a), b), c) and d) separately. 

I propose to handle the complete generate refactor in multiple steps.
1) Define the new `generate()` design only for the simplest use case being `greedy_search` (a)). Leave all existing code for now to not break any of b), c), d)
2) At this point we can already start change greedy search to be XLA-compatible.
3) Analogues to 1), we define the sub-method b) and can then fully remove `_generate_no_beam_search`
4) Refactor c) beam search. This is a bit more complex and I'd like to handle this in a second step.
5) Finally we can refactor d) which should be relatively simple after 4) and can then remove the `_generate_beam_search` function.

This first PR lays the foundation for the complete refactor and finishes 1).

In short in does the following:

- i.) Adds more aggressive `greedy_search` tests for a decoder-only model - GPT2 - and an encoder-decoder model - T5.
- ii.) Define the TF logic for logits processors
- iii.) Define the new `generate()` method which is in a first step called `_generate()` so that it will only be used for a) `greedy_search`. It should be much more readable by adding clear `# 1. ... 2. ` comments for each model input preparation step.
- iv.) Fully refactor `greedy_search`
- v.) Clean-up some duplicated code that was heavily used by generate and related functions (*i.e.* `tf_utils.py`).

## Next steps
=> In a next step, I propose to tackle both 2) and 3) at the same time. 
@Rocketknight1 Maybe you could take care of 2) while maybe @gante maybe it would make sense if you start with 3) (should be pretty straight-forward after this PR). IMO, this would also greatly help at making more people in the team knowledgeable in how `generate()` works.
Before we go over to 4), I'm happy to do the `past`/`encoder_outputs` refactor (see all those `TODO(Patrick)` comments in `generate`). We still rely on a very ugly hack that puts `encoder_outputs` into the `past` variable. I didn't want to change this here because it require changing the `prepare_inputs_for_generation` method of multiple models and IMO the PR would have become to difficult to review.

Thoughts? @Rocketknight1 @gante @LysandreJik @sgugger

### CODE REVIEW

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:1220]:
```suggestion
                # from fairseq: https://github.com/pytorch/fairseq/blob/a07cb6f40480928c9e0548b737aadd36ee66ac76/fairseq/sequence_generator.py#L345
```

**Comment by patrickvonplaten** [comment]
Testing all of the following slow tests:

```bash
RUN_SLOW=1 pytest -sv \
tests/test_modeling_tf_bart.py \
tests/test_modeling_tf_t5.py \
tests/test_modeling_tf_gpt2.py \
tests/test_modeling_tf_vision_encoder_decoder.py \
tests/test_modeling_tf_encoder_decoder.py \
tests/test_modeling_tf_speech_to_text.py
```

as well as

```bash
RUN_SLOW=1 pytest -sv tests/test_modeling_tf_rag.py
```

File: src/transformers/modeling_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:2050]: This was defined in three different places previously. Instead of copy-pasting this function a for a fourth time, I moved it to `tf_utils.py` along with `set_tensor_by_indices_to_value`

Comment (patrickvonplaten) [reply src/transformers/modeling_tf_utils.py:2050]: This is the reason why so many files are changed in this PR. It was quite entangled in the changes of this PR so I hope it's ok to have this refactor be included here.

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:558]: As one can see the original `generate()` method is not really touched in this PR. Here all that is done is to force to use the new `_generate()` method for the greedy search use case. IMO, this can be left as is for now and makes the refactor much easier to split into multiple steps while being sure nothing is broken down the line.

File: src/transformers/generation_tf_utils.py
Comment (LysandreJik) [inline src/transformers/generation_tf_utils.py:1927]: Interesting, why is that needed?

File: src/transformers/generation_tf_logits_process.py
Comment (LysandreJik) [inline src/transformers/generation_tf_logits_process.py:79]: That's quite cool

File: src/transformers/modeling_tf_utils.py
Comment (LysandreJik) [reply src/transformers/modeling_tf_utils.py:2050]: Ok for me, thanks for taking care of it

**Review by LysandreJik — APPROVED** [review]
Thanks for working on it, this is great! The TODOs spread everywhere should make it simpler to continue.

Great test addition.

File: src/transformers/generation_tf_utils.py
Comment (sgugger) [inline src/transformers/generation_tf_utils.py:1790]: This error message could be split in several lines to be visible at 119 char limit  ;-)

File: src/transformers/modeling_tf_utils.py
Comment (sgugger) [reply src/transformers/modeling_tf_utils.py:2050]: I trust you to have moved everything needed along with it. Could you make a good first issue with other functions we could move to that new module? Similarly with `pytorch_utils`, we would only keep the classes for TFPreTrainedModel and models in `modeling_tf_utils` and have all util functions in `tf_utils`.

File: tests/test_modeling_tf_gpt2.py
Comment (sgugger) [inline tests/test_modeling_tf_gpt2.py:513]: Could be put on its own line with a black-ignore comment (and the comment on the right on the line above).

**Review by sgugger — APPROVED** [review]
Amazing work and very clear roadmap! Thanks a lot for working on this!

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [reply src/transformers/generation_tf_utils.py:1927]: It seems like TF 2.8.0 doesn't have a `dtype` input arg anymore: https://www.tensorflow.org/api_docs/python/tf/fill . Not sure why this worked before

File: src/transformers/generation_tf_logits_process.py
Comment (Rocketknight1) [reply src/transformers/generation_tf_logits_process.py:78]: Will this throw an error when a kwarg is only used by some of the `LogitsProcessor`s in the list? Most `LogitsProcessor` classes don't seem to have `**kwargs` in their call, so they will error out if they get an unexpected keyword argument, right?

File: tests/test_generation_tf_logits_process.py
Comment (Rocketknight1) [inline tests/test_generation_tf_logits_process.py:44]: Typo: "length"

File: tests/test_generation_tf_logits_process.py
Comment (gante) [inline tests/test_generation_tf_logits_process.py:126]: I'm looking at this test and I noticed I probably didn't get the processor correctly. If we have `bad_word_tokens=[[99, 100]]`, it means that the pair (`[99, 100]`) is forbidden, but the individual tokens (`99` and `100`) are not, correct?

If what I wrote above is correct, shouldn't the first batch have `(1, 3)` and the second batch `(1, 2, 3)` forbidden? What am I missing?

**Review by gante — APPROVED** [review]
Thanks for this big rework <3

File: src/transformers/tf_utils.py
Comment (Rocketknight1) [inline src/transformers/tf_utils.py:29]: The `value` argument to `tf.where` can broadcast even from Python scalar inputs, so I don't think the `zeros_like` tensor is ever necessary.
```suggestion
    return tf.where(indices, value, tensor)
```

**Review by Rocketknight1 — APPROVED** [review]

**Comment by Rocketknight1** [comment]
Just finished and it looks very cool! I had a few nits but it seems like a great refactor, which will make it easy to optimize sub-methods for performance later.

File: src/transformers/generation_tf_logits_process.py
Comment (patrickvonplaten) [reply src/transformers/generation_tf_logits_process.py:155]: Those logits processor are definitely slow right now and it would be a very welcoming change to speed them up!

File: src/transformers/generation_tf_logits_process.py
Comment (patrickvonplaten) [reply src/transformers/generation_tf_logits_process.py:138]: Actually, I had to add them to make the code work with TF 2.7.0 I think. Is it possible that we need this for older TF versions?

File: tests/test_generation_tf_logits_process.py
Comment (patrickvonplaten) [reply tests/test_generation_tf_logits_process.py:126]: Maybe the comments aren't super clear here, but the logic is the following. For both inputs (batch_1 and batch_2) we want to know which word ids we are allowed to generate for the next step. The logits vector is of dim 5 thus covering the word ids [0-4] . Now the word ids [1], [4] are always forbidden, which is why we have `True` for both output lists in the last place [4th logit] and the second one [1st]. The other three sequences are then responsible for setting the other values (0th and 3rd for the first batch) and (0th and 2nd for the second batch)

File: src/transformers/generation_tf_logits_process.py
Comment (patrickvonplaten) [reply src/transformers/generation_tf_logits_process.py]: It should throw an error if a logits processor requires more kwargs than the usual two (`input_ids` and `scores`) but it is not given to the list. Might be a bit overkill this check, but shouldn't hurt. You reminded me that the 3 is incorrect there though and that it should be a two :-)

**Comment by patrickvonplaten** [comment]
Thanks a lot for the review @gante and @Rocketknight1 . Adapted the PR according to your comments and will merge once tests are passing.

File: tests/test_generation_tf_logits_process.py
Comment (gante) [reply tests/test_generation_tf_logits_process.py:126]: Gotcha -- e.g. because both batches end with a `1` and `[1, 0]` is a bad sequence, generating a `0` in the next step is forbidden. Thanks 👍

File: src/transformers/generation_tf_logits_process.py
Comment (Rocketknight1) [reply src/transformers/generation_tf_logits_process.py]: I'm unsure though! Let's say we have two `LogitsProcessor`s in our list. `processor_1` takes `input_ids` and `scores` and `processor_2` takes `input_ids`, `scores` and `extra_data`. Then if we pass an `extra_data` argument to `TFLogitsProcessorList.__call__`, it will crash when it calls `processor_1` because `processor_1` has no `extra_data` or `**kwargs` argument. If we don't pass `extra_data` to `TFLogitsProcessorList.__call__` then it will crash when it calls `processor_2` because the `extra_data` argument for `processor_2` will be missing.

It seems like all `LogitsProcessor` classes need a `**kwargs` argument in their `__call__` method to catch extra unused arguments, or else they need to all have the same `__call__` signature.

**Comment by patrickvonplaten** [comment]
Failing tests are flaky and unrelated. Merging

File: src/transformers/generation_tf_logits_process.py
Comment (patrickvonplaten) [reply src/transformers/generation_tf_logits_process.py]: Yeah not really sure what was going on, but I now removed all `.cpu()` statements and re-tested it. It seems to work fine now!

**Comment by ydshieh** [comment]
@patrickvonplaten 

About the following

```
We still rely on a very ugly hack that puts encoder_outputs into the past variable. I didn't want to change this here because it require changing the prepare_inputs_for_generation method of multiple models and IMO the PR would have become to difficult to review.
```

I haven't looked the details, but it is something different from the PyTorch's models (and `generate`), right?
During the fix for PT/TF inconsistency, this causes some problem for a strict equivalence testing.
Maybe I can work on this once other inconsistencies are fixed (and by that time, `generate` already has some clean refactoring by you guys!).

**Comment by patrickvonplaten** [comment]
Hey @ydshieh, 

Great initiative! Yes that's indeed an important refactor we should do soon - would you like to look into it? In short we should do the following in TF:

- Always force `return_dict=True`. The we don't need any of these `self._use_cache()` methods anymore (we can just check whether `past_key_values` is present in the output. Also this way the can clearly differentiate between `encoder_outputs` and `past_key_values` for the encoder-decoder case. Would you like to open a PR for it? Happy to help you along the way!

**Comment by ydshieh** [comment]
> Hey @ydshieh,
> 
> Great initiative! Yes that's indeed an important refactor we should do soon - would you like to look into it? In short we should do the following in TF:
> 
>     * Always force `return_dict=True`. The we don't need any of these `self._use_cache()` methods anymore (we can just check whether `past_key_values` is present in the output. Also this way the can clearly differentiate between `encoder_outputs` and `past_key_values` for the encoder-decoder case. Would you like to open a PR for it? Happy to help you along the way!

Sure - after I finish the work of adding testing `past_key_values` for TF encoder models (those can be used for causal lm and can have cross-attention`). Currently, no test for it, and making #15477 undetected for quite some time.

File: tests/test_modeling_tf_gpt2.py
Comment (patrickvonplaten) [inline tests/test_modeling_tf_gpt2.py:445]: add a new agressive test

Comment (Goga345345) [reply tests/test_modeling_tf_gpt2.py:445]: Whwreto buy proper pyros?

**Pull request merged**
